### PR TITLE
Feature etb#240 tf luna lidar driver

### PR DESCRIPTION
**Pull request by jrachucki-n7s**

### CODE REVIEW

File: Drivers/c/LidarExample/main.c [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+                    &prvTaskBuffer);
+
+  vTaskStartScheduler();
+  for (;;) {
+  }
+}
+
+static void parseTfLunaData() {
+  uint8_t rxByte;
+  uint8_t checksum;
```
Comment (Lurkerpas) [inline Drivers/c/LidarExample/main.c]: I think that reliance on unsigned int 8 overflow for modulo is not a great idea. Could we change this to e.g. uint32_t and then just do %?

File: Drivers/c/LidarExample/main.c [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+
+static void prvTask(void *pvParameters);
+
+void lidarRxCallback(uint8_t readByte) {
+  if (!ByteFifo_push(&rxFifo, readByte)) {
+    Hal_console_usart_write("Fifo Full\n", 10);
+  }
+}
+
+void lidarTxCallback(void) {
```
Comment (Lurkerpas) [inline Drivers/c/LidarExample/main.c]: Is it used?

File: Drivers/c/LidarExample/main.c [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+
+  /* Block for 500ms. */
+  const TickType_t xDelay = 500 / portTICK_PERIOD_MS;
+
+  ByteFifo_init(&rxFifo, lidarRxBuffer, LIDAR_RX_BUFFER_LEN);
+  UartHwas_init(&uart, &config);
+  UartHwas_readByteAsync(&uart, lidarRxCallback);
+
+  for (;;) {
+    parseTfLunaData();
```
Comment (Lurkerpas) [inline Drivers/c/LidarExample/main.c:141]: Could we drive parseIfLunaData by lidarRxCallback, e.g., via event or message?

File: Drivers/seds/AfecHwas.xml [hunk 4]
```diff
@@ -0,0 +1,84 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
```
Comment (Lurkerpas) [inline Drivers/seds/AfecHwas.xml:1]: Why is this modified in the scope of this MR?

File: Drivers/seds/UartHwasComponent.xml [hunk 5]
```diff
@@ -0,0 +1,1251 @@
...
+              </Body>
+            </Activity>
+            <Activity name="SendByteAsync">
+              <Argument name="uart" type="com/n7space/UartHwas/UartHwas"></Argument>
+              <Argument name="byte" type="com/n7space/Hwas/Byte"></Argument>
+              <Body>
+                <Assignment outputVariableRef="irqNumberVar">
+                  <VariableRef variableRef="uartVar.mIrqNumber" />
+                </Assignment>
+                <SendCommandPrimitive shortDescription="Subscribe to Uart interrupt" interface="InterruptSubscriptionManagement" command="SubscribeToInterrupt">
```
Comment (Lurkerpas) [inline Drivers/seds/UartHwasComponent.xml]: Are the changes from the previous MR propagated to this one?

File: Drivers/seds/UartHwasExample/linuxPartition/main.cpp [hunk 6]
```diff
@@ -0,0 +1,61 @@
+/**@file
```
Comment (Lurkerpas) [inline Drivers/seds/UartHwasExample/linuxPartition/main.cpp]: I remember looking at this file.... Are you sure that everything is properly updated and targetted?

File: Drivers/seds/TfLunaHwas.xml [hunk 7]
```diff
@@ -0,0 +1,61 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<PackageFile xmlns="http://www.ccsds.org/schema/sois/seds" xmlns:xi="http://www.w3.org/2001/XInclude" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.ccsds.org/schema/sois/seds seds.xsd">
+  <Package name="com/n7space/TfLunaHwas">
```
Comment (Lurkerpas) [inline Drivers/seds/TfLunaHwas.xml]: Pio, Afec and Uart use Hwas.... Luna does not, right? So maybe we could drop Hwas from the name?

File: Drivers/seds/TfLunaHwas.xml [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+<PackageFile xmlns="http://www.ccsds.org/schema/sois/seds" xmlns:xi="http://www.w3.org/2001/XInclude" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.ccsds.org/schema/sois/seds seds.xsd">
+  <Package name="com/n7space/TfLunaHwas">
+    <DataTypeSet>
+      <EnumeratedDataType name="TfLunaHwas_ReadSM">
+        <LongDescription>Enum listing UART read data frame state machine states</LongDescription>
+        <IntegerDataEncoding encoding="unsigned" byteOrder="littleEndian" sizeInBits="8" />
+        <EnumerationList>
+          <Enumeration label="TfLunaHwas_ReadSM_NotFound" value="0" />
+          <Enumeration label="TfLunaHwas_ReadSM_Head" value="1"></Enumeration>
+          <Enumeration label="TfLunaHwas_ReadSM_DisL" value="3" />
```
Comment (Lurkerpas) [inline Drivers/seds/TfLunaHwas.xml]: DistanceL, or DistanceLow/Lower?

File: Drivers/seds/TfLunaHwas.xml [hunk 9]
```diff
@@ -0,0 +1,61 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<PackageFile xmlns="http://www.ccsds.org/schema/sois/seds" xmlns:xi="http://www.w3.org/2001/XInclude" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.ccsds.org/schema/sois/seds seds.xsd">
+  <Package name="com/n7space/TfLunaHwas">
+    <DataTypeSet>
+      <EnumeratedDataType name="TfLunaHwas_ReadSM">
+        <LongDescription>Enum listing UART read data frame state machine states</LongDescription>
+        <IntegerDataEncoding encoding="unsigned" byteOrder="littleEndian" sizeInBits="8" />
+        <EnumerationList>
+          <Enumeration label="TfLunaHwas_ReadSM_NotFound" value="0" />
+          <Enumeration label="TfLunaHwas_ReadSM_Head" value="1"></Enumeration>
```
Comment (Lurkerpas) [inline Drivers/seds/TfLunaHwas.xml]: Header?

File: Drivers/seds/TfLunaHwas.xml (hunk 8 above)
Comment (Lurkerpas) [reply Drivers/seds/TfLunaHwas.xml]: Also for Temperature and Strength...

File: Drivers/seds/TfLunaHwas.xml [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+          <Entry name="mDistance" type="com/n7space/Hwas/Word" />
+          <Entry name="mStrength" type="com/n7space/Hwas/Word" />
+          <Entry name="mTemp" type="com/n7space/Hwas/Word" />
+        </EntryList>
+      </ContainerDataType>
+      <EnumeratedDataType name="TfLunaHwas_DataFrameErr">
+        <LongDescription>Enum listing possible rx errors</LongDescription>
+        <IntegerDataEncoding encoding="unsigned" byteOrder="littleEndian" sizeInBits="8" />
+        <EnumerationList>
+          <Enumeration label="TfLunaHwas_DataFrameErr_NoHead1" value="0" />
```
Comment (Lurkerpas) [inline Drivers/seds/TfLunaHwas.xml]: OK... So is this header, or just head (I would be suprised, but if this is in the Luna's docs....)?

File: Drivers/seds/TfLunaHwas.xml [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+          <Enumeration label="TfLunaHwas_DataFrameErr_NoHead2" value="1" />
```
Comment (Lurkerpas) [inline Drivers/seds/TfLunaHwas.xml]: What is the difference brtween NoHead1 and NoHead2?

File: Drivers/seds/TfLunaHwas.xml [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+      <Interface level="functional" name="TfLunaHwasInterfaceType">
+        <CommandSet>
+          <Command name="InitTfLunaCmd" mode="sync">
+            <LongDescription>Initializes uart for communication with LIDAR</LongDescription>
+          </Command>
+        </CommandSet>
+      </Interface>
+      <Interface level="functional" name="TfLunaReturnHwasInterfaceType">
+        <CommandSet>
+          <Command name="ReadDataCmd" mode="sync">
```
Comment (Lurkerpas) [inline Drivers/seds/TfLunaHwas.xml]: Would it be an issue to have these as async commands? I think it would be easier for me to work with asyncs in SDL further down the project.

File: Drivers/seds/TfLunaHwasComponent.xml [hunk 13]
```diff
@@ -0,0 +1,390 @@
...
+                  <ArgumentValue name="uart">
+                    <VariableRef variableRef="uartVar" />
+                  </ArgumentValue>
+                  <ArgumentValue name="config">
+                    <VariableRef variableRef="uartConfigVar" />
+                  </ArgumentValue>
+                </SendCommandPrimitive>
+                <Assignment outputVariableRef="readStateVar">
+                  <LongDescription>Init state machine</LongDescription>
+                  <Value value="TfLunaHwas_ReadSM_NotFound" />
```
Comment (Lurkerpas) [inline Drivers/seds/TfLunaHwasComponent.xml]: Maybe this state could be called Idle, or Wait/Waiting? NotFound might be misleading

File: Drivers/seds/TfLunaHwasComponent.xml [hunk 14]
```diff
@@ -0,0 +1,390 @@
...
+                </Assignment>
+              </Body>
+            </Activity>
+            <Activity name="IncreaseCheckSum">
+              <Argument name="byte" type="com/n7space/Hwas/Word" />
+              <Body>
+                <MathOperation outputVariableRef="checksumVar">
+                  <Operator operator="add" />
+                  <VariableRef variableRef="checksumVar" />
+                  <VariableRef variableRef="byte" />
```
Comment (Lurkerpas) [inline Drivers/seds/TfLunaHwasComponent.xml]: Where is the modulo operation?

File: Drivers/seds/TfLunaHwasComponent.xml (hunk 13 above)
Comment (jrachucki-n7s) [reply Drivers/seds/TfLunaHwasComponent.xml]: Sure, I will change that

File: Drivers/seds/TfLunaHwas.xml (hunk 12 above)
Comment (jrachucki-n7s) [reply Drivers/seds/TfLunaHwas.xml]: I wouldn't be an issue. I will change that.

File: Drivers/seds/TfLunaHwas.xml (hunk 11 above)
Comment (Lurkerpas) [reply Drivers/seds/TfLunaHwas.xml]: Ok... Maybe this should be IncorrectHeaderByte1, IncorrectHeaderByte2, or IncorrectFirstHeaderByte, IncorrectSecondHeaderByte. Just an idea...

Comment (jrachucki-n7s) [reply Drivers/seds/TfLunaHwas.xml]: NoHead1 means that we didn't received first character of the header. So, assuming that we have enabled the interrupt in the middile of the lidar message, for every character that we have received we will get a frame error NoHead1. 
NoHead2 means that we have received first character of the header, but the character received after that is not correct. It might indicate that at least one byte was lost. The header looks like this 0x59 0x59. After that there are three 16 bits integers and 8 bit crc.

File: Drivers/seds/TfLunaHwas.xml (hunk 10 above)
Comment (jrachucki-n7s) [reply Drivers/seds/TfLunaHwas.xml]: Yeah, ok. It should be called a header. Nevertheless, the documentation uses a word head, referring to the message header. (but not in the case of the first two bytes of that particular message, refer to page 24 of this manual https://download.kamami.pl/p579517-15978504093169389SJ-PM-TF-Luna%20A05%20Product%20Manual.pdf)

File: Drivers/seds/UartHwasExample/linuxPartition/main.cpp (hunk 6 above)
Comment (jrachucki-n7s) [reply Drivers/seds/UartHwasExample/linuxPartition/main.cpp]: I will check that. 
Here it should be indicated that this file was renamed, not created.

File: Drivers/c/LidarExample/main.c (hunk 2 above)
Comment (jrachucki-n7s) [reply Drivers/c/LidarExample/main.c]: No, I will delete that.

File: Drivers/c/LidarExample/main.c (hunk 1 above)
Comment (jrachucki-n7s) [reply Drivers/c/LidarExample/main.c]: Sure, I will change that.

File: Drivers/c/LidarExample/main.c (hunk 3 above)
Comment (jrachucki-n7s) [reply Drivers/c/LidarExample/main.c:141]: I think so. I will try to change it.

File: Drivers/seds/UartHwasComponent.xml (hunk 5 above)
Comment (jrachucki-n7s) [reply Drivers/seds/UartHwasComponent.xml]: The name of this file has changed, and it was split into two files. I guess that's why changes were not propagated. I will try to sync this branch with the master.

File: Drivers/seds/AfecHwas.xml (hunk 4 above)
Comment (jrachucki-n7s) [reply Drivers/seds/AfecHwas.xml:1]: Besides implementing lidar, I have also split every SEDS file into interface and implementation. I can try to isolate this change into a separate branch if you like.

Comment (Lurkerpas) [reply Drivers/seds/AfecHwas.xml:1]: OK, so this is just this split? I will review it in more detail then.

File: Drivers/seds/TfLunaHwas.xml (hunk 10 above)
Comment (Lurkerpas) [reply Drivers/seds/TfLunaHwas.xml]: OK, so if the docs say Head, let's stay with Head.

File: Drivers/seds/TfLunaHwas.xml (hunk 11 above)
Comment (Lurkerpas) [reply Drivers/seds/TfLunaHwas.xml]: Undestood. Let's stay with Head instead of Header, but still, IncorrectFirstHeadByte would be IMHO much more readable.

File: Drivers/seds/TfLunaHwas.xml (hunk 9 above)
Comment (Lurkerpas) [reply Drivers/seds/TfLunaHwas.xml]: Forget it, Head is OK.

File: Drivers/seds/TfLunaHwasComponent.xml (hunk 14 above)
Comment (jrachucki-n7s) [reply Drivers/seds/TfLunaHwasComponent.xml]: I just realized that this SEDS is not the newest version. The SDL is correct, though. It happened because the SDLs are manually generated locally on my VM and then copied here. It is done like this as TASTE, as far as I know, doesn't support that yet. 
I will fix that

Comment (jrachucki-n7s) [reply Drivers/seds/TfLunaHwasComponent.xml]: Done. The newest version was pushed. Although modulo is calculated like this: CRC = (A+B) - (floor((A+B)/256) * 256.

File: Drivers/seds/TfLunaHwasComponent.xml (hunk 13 above)
Comment (jrachucki-n7s) [reply Drivers/seds/TfLunaHwasComponent.xml]: Done

File: Drivers/seds/TfLunaHwas.xml (hunk 8 above)
Comment (jrachucki-n7s) [reply Drivers/seds/TfLunaHwas.xml]: Done

File: Drivers/seds/TfLunaHwas.xml (hunk 7 above)
Comment (jrachucki-n7s) [reply Drivers/seds/TfLunaHwas.xml]: done

File: Drivers/seds/UartHwasExample/linuxPartition/main.cpp (hunk 6 above)
Comment (jrachucki-n7s) [reply Drivers/seds/UartHwasExample/linuxPartition/main.cpp]: done

File: Drivers/seds/UartHwasComponent.xml (hunk 5 above)
Comment (jrachucki-n7s) [reply Drivers/seds/UartHwasComponent.xml]: done

File: Drivers/c/LidarExample/main.c (hunk 3 above)
Comment (Lurkerpas) [reply Drivers/c/LidarExample/main.c:141]: Not changed, but as we have working SEDS, this is not that important. Let's prioritize other tasks.

File: Drivers/c/LidarExample/main.c (hunk 2 above)
Comment (Lurkerpas) [reply Drivers/c/LidarExample/main.c]: Still unused, but OK, let's prioritize other issues.

File: Drivers/seds/TfLunaHwas.xml (hunk 12 above)
Comment (Lurkerpas) [reply Drivers/seds/TfLunaHwas.xml]: This is thill sync....

Comment (Lurkerpas) [reply Drivers/seds/TfLunaHwas.xml]: OK, after discussion, this command can be "asynced" later in the chain.

**Review by Lurkerpas — APPROVED** [review]
